### PR TITLE
Enhance subgraph API

### PR DESCRIPTION
**Pull request by ZhennanQin**
@reminisce @zheng-da @apeforest @szha  @pengzhao-intel
PR Major change:

* Introduce `SubgraphSelectorV2` that accepts `BiDirectedNode` instead of `nnvm::Node` to get full context of visited node. `BiDirectedNode` is double linked version of  `nnvm::Node`, which can list the output consumer of node. This is very useful to help make selection which has output requirement. Wrapper class `SubgraphSelectorV2Bridge` is added to convert `SubgraphSelector` to `SubgraphSelectorV2`, which allows developer to continue using old `SubgraphSelector`. So this feature is fully back-compatible.

* Introduce `kAdjust` mode for `SubgraphProperty` to allow adjusting network parameters instead of create new node. New API `AdjustSubgraphNode()` is added to  `SubgraphProperty` for this, which can adjust selected nodes attributes. This is useful when we don't want to create new subgraph node, but to do some modification on selected nodes. Quantized scale alignment optimization is pending on this feature.

* Change the behavior of PartitionGraph to create single subgraph node for nodes isolating with each other. https://cwiki.apache.org/confluence/display/MXNET/Enable+Operator+Level+Parallelism+under+Subgraph is depending on this change.

* Add new subgraph pass "quantized scale alignment optimization" as an example for new subgraph API.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by pengzhao-intel** [comment]
@KellenSunderland @TaoLv

**Comment by vandanavk** [comment]
@mxnet-label-bot add [pr-awaiting-review, Operator]

**Comment by vandanavk** [comment]
@ZhennanQin could you re-trigger the CI?

**Comment by karan6181** [comment]
@ZhennanQin Could you please re-trigger the CI and see why some of the checks are still waiting? Thanks !

**Comment by pengzhao-intel** [comment]
@anirudh2290 @reminisce please help review the continuous improvements for subgraph after #14277

File: src/operator/subgraph/build_subgraph.cc [hunk 1]
```diff
@@ -101,12 +68,12 @@ void PrintNodeEntries(const std::vector<nnvm::NodeEntry*>& entries) {
  * \param g the MXNet computational graph
  * \param simple_nodes the nodes of undirected graph in top sorted order
  */
-void CreateSimpleGraph(const Graph& g,
-                       std::vector<SimpleNodePtr>* simple_nodes) {
+void CreateSimpleGraph(const nnvm::Graph& g,
+                       std::vector<BiDirectionalNodePtr>* simple_nodes) {
```
Comment (reminisce) [inline src/operator/subgraph/build_subgraph.cc]: BiDirectional -> BiDirected is more appropriate.

File: src/operator/subgraph/build_subgraph.cc [hunk 2]
```diff
@@ -153,20 +120,23 @@ void ResetNodeLabels(const nnvm::Graph& g,
...
-  }
+  std::queue<BiDirectionalNode*> node_queue;
+  CHECK_EQ(simple_nodes[snid]->label, -1);
+  simple_nodes[snid]->label = label;
+  node_queue.push(simple_nodes[snid].get());
+  struct hash_pair {
+    inline size_t operator()(
+        const std::pair<const BiDirectionalNode*, const BiDirectionalNode*>& v) const {
+      size_t value = std::hash<const void*>()(v.first);
+      value ^= std::hash<const void*>()(v.second) + 0x9e3779b9 + (value << 6) + (value >> 2);
```
Comment (reminisce) [inline src/operator/subgraph/build_subgraph.cc]: Use `dmlc::HashCombine` to get the hash value of a pair of nodes.

File: src/operator/subgraph/build_subgraph.cc [hunk 3]
```diff
@@ -180,41 +150,40 @@ bool LabelSubgraph(const Graph& g,
     std::pair<std::vector<const nnvm::Node*>,
               std::vector<const nnvm::Node*>>> non_subgraph_node_map;
   while (!node_queue.empty()) {
-    SimpleNode* cur_node = node_queue.front();
+    auto cur_node = node_queue.front();
```
Comment (reminisce) [inline src/operator/subgraph/build_subgraph.cc]: nit: Be expressive by avoiding using `auto`. `auto` is not good for readability.

File: src/operator/subgraph/build_subgraph.cc [hunk 4]
```diff
@@ -676,14 +605,30 @@ void CreateSubgraphNode(Graph* g,
 #endif
 }
 
+/*!
+ * \brief Adjust a set of nodes belonging to the same subgraph.
```
Comment (reminisce) [inline src/operator/subgraph/build_subgraph.cc]: Please elaborate the meaning of `adjust`. Maybe using an example is good for explanation.

File: src/operator/subgraph/build_subgraph.cc [hunk 5]
```diff
@@ -748,27 +693,38 @@ Graph PartitionGraph(Graph&& g) {
...
-    CreateSubgraphNode(&g, simple_nodes, subgraph_nodes[i], i, &entry_top_order_map);
-  }
+    auto ptype = subg_prop->GetPropertyType();
+    if (ptype == SubgraphProperty::SgPropertyType::kCreate) {
+      CreateSubgraphNode(&g, simple_nodes, subgraph_nodes[i], subgraph_selectors[i], i,
+                         &entry_top_order_map);
+    } else {
+      CHECK_EQ(ptype, SubgraphProperty::SgPropertyType::kAdjust);
+      AdjustSubgraphNode(&g, subgraph_nodes[i], subgraph_selectors[i], i);
+      }
```
Comment (reminisce) [inline src/operator/subgraph/build_subgraph.cc]: Indentation seems wrong.

File: src/operator/subgraph/mkldnn/mkldnn_post_quantize_align_scale_property.h [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+  static SubgraphPropertyPtr Create() {
+    static const std::string &name = "MKLDNN post-quantization scale alignment optimization pass";
+    auto property = std::make_shared<SgMKLDNNPostQuantizeAlignScaleProperty>();
+    property->SetAttr<std::string>("property_name", name);
+    property->SetAttr<bool>("inference_only", true);
+    return property;
+  }
+
+  void AdjustSubgraphNode(const std::vector<nnvm::Node *> &subgraph_nodes,
```
Comment (reminisce) [inline src/operator/subgraph/mkldnn/mkldnn_post_quantize_align_scale_property.h]: Please add docstring for this function.

**Comment by pengzhao-intel** [comment]
@zhennanqin @xinyu-intel please resolve the comments and re-trigger the PR.
We need to speedup the progress of this changes :)

File: src/operator/subgraph/build_subgraph.cc (hunk 1 above)
Comment (ZhennanQin) [reply src/operator/subgraph/build_subgraph.cc]: Changed.

File: src/operator/subgraph/build_subgraph.cc (hunk 2 above)
Comment (ZhennanQin) [reply src/operator/subgraph/build_subgraph.cc]: Seems this is redundant code. Removed.

File: src/operator/subgraph/build_subgraph.cc (hunk 4 above)
Comment (ZhennanQin) [reply src/operator/subgraph/build_subgraph.cc]: Added.

File: src/operator/subgraph/build_subgraph.cc (hunk 5 above)
Comment (ZhennanQin) [reply src/operator/subgraph/build_subgraph.cc]: Changed.

File: src/operator/subgraph/mkldnn/mkldnn_post_quantize_align_scale_property.h (hunk 6 above)
Comment (ZhennanQin) [reply src/operator/subgraph/mkldnn/mkldnn_post_quantize_align_scale_property.h]: Added.

File: src/operator/subgraph/build_subgraph.cc (hunk 3 above)
Comment (ZhennanQin) [reply src/operator/subgraph/build_subgraph.cc]: Changed.

**Comment by pengzhao-intel** [comment]
Thanks for your contribution. Merging now.

**Pull request merged**